### PR TITLE
improve str macro doc help

### LINKED ISSUE #38334: improve help search for string macros
If you want to know about e.g. the raw string macro, `?raw` does successfully turn up `@raw_str`. However, that is a bit obscure, so we should perhaps include `raw""` in the results as well.

### PR DESCRIPTION
**Pull request by Moelf**
close #38334
before:
```
help?> raw
search: RawFD @raw_str RoundNearestTiesAway

Couldn't find raw
Perhaps you meant rand, read, real, rpad, rem, rm, run, NaN, Val, cat, map, max, tan, new, RawFD, ratio, randn or range
  No documentation found
```

after:
```
help?> raw
search: RawFD raw"" @raw_str RoundNearestTiesAway

Couldn't find raw
Perhaps you meant rand, read, real, rpad, rem, rm, run, NaN, Val, cat, map, max, tan, new, RawFD, randn, range or read!
  No documentation found.
```

### CODE REVIEW

File: stdlib/REPL/src/docview.jl [hunk 1]
```diff
@@ -645,7 +645,17 @@ accessible(mod::Module) =
...
-doc_completions(name) = fuzzysort(name, accessible(Main))
+function doc_completions(name) 
+    res = fuzzysort(name, accessible(Main))
+
+    # to insert an entry like `raw""` for `"@raw_str"` in `res`
+    ms = match.(r"^@(.*?)_str$", res)
+    idxs = findall(!isnothing, ms)
+    for i in idxs
+        insert!(res, i, "$(only(ms[i].captures))\"\"")
+    end
```
Comment (NHDaly) [inline stdlib/REPL/src/docview.jl:656]: After your first insertion, I think all of the subsequent indexes will be off-by-one, and then after the next, off-by-two, etc.

You can avoid this by inserting in reverse order:
```suggestion
    for i in reverse(idxs)
        insert!(res, i, "$(only(ms[i].captures))\"\"")
    end
```

Comment (NHDaly) [reply stdlib/REPL/src/docview.jl:656]: (Here's an example of how the behavior in the PR so far):
```julia
julia> res = ["a", "@b_str", "@c_str", "d", "e", "@f_str"];

julia> begin
           # to insert an entry like `raw""` for `"@raw_str"` in `res`
           ms = match.(r"^@(.*?)_str$", res)
           idxs = findall(!isnothing, ms)
           for i in idxs
               insert!(res, i, "$(only(ms[i].captures))\"\"")
           end
       end

julia> show(res)
["a", "b\"\"", "c\"\"", "@b_str", "@c_str", "f\"\"", "d", "e", "@f_str"]
```

and with `reverse`, they insert in the right spot:
```julia
julia> res = ["a", "@b_str", "@c_str", "d", "e", "@f_str"];

julia> begin
           # to insert an entry like `raw""` for `"@raw_str"` in `res`
           ms = match.(r"^@(.*?)_str$", res)
           idxs = findall(!isnothing, ms)
           for i in reverse(idxs)
               insert!(res, i, "$(only(ms[i].captures))\"\"")
           end
       end

julia> show(res)
["a", "b\"\"", "@b_str", "c\"\"", "@c_str", "d", "e", "f\"\"", "@f_str"]
```

**Review by NHDaly — APPROVED** [review]
Looks great to me! :) thanks for the PR!

File: stdlib/REPL/src/docview.jl (hunk 1 above)
Comment (Moelf) [reply stdlib/REPL/src/docview.jl:656]: ah yes, good point, need to insert in the opposite side of the look up direction

**Comment by NHDaly** [comment]
Some of your tests are failing. Also, it would help if you updated the description comment on this PR with an indication of what this changes, and included a before/after example output from the help system.

**Comment by NHDaly** [comment]
Ah, the failing test should be easy to fix.

I copied this from the buildbot links above:
```julia
Error in testset REPL:
Test Failed at /Users/julia/buildbot/worker-tabularasa/tester_macos64/build/share/julia/stdlib/v1.7/REPL/test/replcompletions.jl:10
  Expression: (REPL.doc_completions(i))[1] == i
   Evaluated: "var\"\"" == "@var_str"
```
And here's the test it's referring to:
https://github.com/JuliaLang/julia/blob/f425ce7d440bc9da7361553e9894b122679740fa/stdlib/REPL/test/replcompletions.jl#L8-L12

That last test will fail, since when you ask for `doc_completions("@var_str")`, the first result is now `var""`. You could probably resolve this by changing that line to:
```julia
         @test i ∈ REPL.doc_completions(i)
```
And you might also want to add `var""` to the list of search strings?

**Comment by Moelf** [comment]
> And you might also want to add `var""` to the list of search strings?

what does this mean?

**Comment by ViralBShah** [comment]
Let's rebase this to prepare for merging.

**Comment by NHDaly** [comment]
> > And you might also want to add `var""` to the list of search strings?
> 
> what does this mean?

sorry for my looooong delay. i think that i must have misunderstood something - my concern is addressed. ✅

**Comment by ViralBShah** [comment]
Merge?

**Comment by ViralBShah** [comment]
Merge?

**Comment by Moelf** [comment]
back port 1.8?

File: stdlib/REPL/src/docview.jl [hunk 2]
```diff
@@ -720,7 +720,19 @@ accessible(mod::Module) =
            map(names, moduleusings(mod))...;
            collect(keys(Base.Docs.keywords))] |> unique |> filtervalid
 
-doc_completions(name) = fuzzysort(name, accessible(Main))
+function doc_completions(name)
```
Comment (giordano) [inline stdlib/REPL/src/docview.jl:723]: Ugh, this line had a trailing whitespace

Comment (Moelf) [reply stdlib/REPL/src/docview.jl:723]: Oops

**Pull request merged**